### ISSUE TITLE
Submission 7 (Project/Short)

### ISSUE THREAD

**Issue by opencert** [opencert/workshop-2019#7]
Title: Open and Interactive Learning Resources for Algorithmic Problem Solving

Abstract:
Algorithmic problem solving is a way of approaching and solving problems by using the advances that have been made in the principles of correct-by-construction algorithm design. The approach has been taught at first-year undergraduate level since September 2003 and, since then, a substantial amount of learning materials have been developed.  However, the existing materials are distributed in a conventional and static way (e.g. as a textbook and as several documents in PDF format available online), not leveraging the capabilities provided by modern collaborative and open-source platforms. 
In this paper, we propose the creation of an online, open-source repository of interactive learning materials on algorithmic problem solving. We show how the existing framework Mathigon can be used to support such a repository. By being open and hosted on a platform such as GitHub, the repository enables collaboration and anyone can create and submit new material. Furthermore, by making the material interactive, we hope to encourage engagement with and a better understanding of the materials.
[OpenCERT_2019_paper_7.pdf](https://github.com/opencert/workshop-2019/files/3409547/OpenCERT_2019_paper_7.pdf)

**Comment by milpaolo** [comment]
I will review this paper
Paolo Milazzo

**Comment by paddykrishnan** [comment]
I can also review this paper.
Paddy Krishnan

**Comment by stever00** [comment]
I will review this paper
Steve Reeves

**Comment by mhmartinho** [comment]
I will review this paper.
M. Helena Martinho

**Comment by opencert** [comment]
Assigned PC members:
- Paolo Milazzo
- Steve Reeves
- M. Helena Martinho
- Marcus Specht

**Comment by stever00** [comment]
I don't think you said int he paper whether there is yet any data on: (1) how effective your ideas have turned out so far; (2) how acceptable students find this approach; (3) how well-used this approach is (or might be). Can you say anything about these aspects?

**Comment by jff** [comment]
Hi Steve, thanks for reading the paper and for the questions. We are unsure
whether you are referring to the interactive learning material or to the more
general approach of teaching APS.

Regarding the interactive learning material, we had many teaching sessions
where we presented these topics and we had some quizzes during some sessions.
What we are proposing to create is based on those experiences. However, we have
not yet tested these "_Mathigon-based online interactive materials_" with
students.

Our submission is a project paper (short paper). It describes a new open
community project, whose ideas are at an early stage of development and which
have not yet been throughly evaluated (the problems, puzzles, techniques, and
approach have been evaluated to some extent, but the online Mathigon-based
interactive material has not).

The idea came from discussions with colleagues and students and from
the observation that what most students enjoy during classes is the
interactivity that problem solving offers. When they go through the material on
their own, they loose the interactivity --- this project, which is still in its
embrionary stage, is a proposal to improve that. Our goal with this submission
is to present the idea to the community and obtain feedback so that we can
improve what we have planned for the next steps of this project.


Regarding the more general approach of teaching APS:

  1. We would say that it depends on how you measure 'effectiveness'. The
   approach is effective in the sense that the techniques and methods taught have
   been used to find new solutions to non-trivial problems (e.g. the material 
   in number theory in [5], and the new class of solitaire games 'Turing Tapes' 
   that is published in [MPC'10](https://link.springer.com/chapter/10.1007/978-3-642-13321-3_1)/[SCP'13](https://www.sciencedirect.com/science/article/pii/S0167642312001359) and mentioned in [1]). 
   If by effectiveness you mean students becoming better problem solvers, 
   then it's more difficult to answer. In [6], we did a small-scale experiment 
   to try to determine if a cohort of students became better problem solvers. 
   The focus was the calculational method and, generally, they adopted what 
   was taught in the module for solving problems, but their use of the 
   calculational style was not effective. In [9], we discuss an experiment 
   where pre-university students, after being exposed to some material, were 
   able to apply techniques like invariants by themselves. 

 2. Our experience is that students appreciate the approach. We have presented
 some feedback to a session we delivered on analysing an algorithmic card trick
 in [7]. In [9], feedback from pre-university students can be found --- it
 shows that the material (or, at least, parts of the material) can even be
 taught at pre-university level.
 During some teaching sessions, we have also collected feedback from our
 students. In general, they appreciate the use of recreational problems and the
 interactivity that arises from solving those problems.

 3. This material was taught for about 10 years at the University of Nottingham
 (it started in 2003). It was taught at Teesside University from 2011 to 2018.
 Thousands of students were exposed to the material. Backhouse's book on APS is
 used in the [course CS2104 at Virginia Tech](http://courses.cs.vt.edu/cs2104).
 We believe that creating an interactive APS book will help and increase the
 adoption of this approach to teach APS.

**Comment by opencert** [comment]
Camera-ready version for the post-proceedings.
[paper_7.pdf](https://github.com/opencert/workshop-2019/files/3918594/paper_7.pdf)
